src/lib/operators/abstract_operator.hpp

```diff
@@ -124,6 +124,9 @@ class AbstractOperator : public std::enable_shared_from_this<AbstractOperator>,
   // Set parameters (AllParameterVariants or CorrelatedParameterExpressions) to their respective values
   void set_parameters(const std::unordered_map<ParameterID, AllTypeVariant>& parameters);
 
+  // LQP node with which this operator has been created. Might be uninitialized.
+  std::shared_ptr<const AbstractLQPNode> lqp_node;
```

Comment:
Ok, here's an idea on this whole circular dependency thing: How about a test where you create a pipeline, retrieve a weak_ptr (!) to one of the LQP nodes, let the pipeline expire, and assert that the weak_ptr is expired?

Reply:
I don't understand what this would actually test. :-/

Reply:
If we had circular ownerships (leading to memory leaks), the use_count of the shared_ptrs would never reach 0. By holding a weak_ptr, we can check if the LQP and PQP nodes are properly destroyed once their pipeline and cache entries are gone.

Reply:
So I keep the abstract operator as is and just test, that the shared_ptr to the LQP is nicely cleaned up. With this test, which should expectedly break right now, we can later see if there are new references which keep the LQP nodes alive. And thus the expected throw no longer occurs.

Do I get this right?

Reply:
Why would it break now? The PQP operator should be removed at some point at which the use_count of the LQP becomes 0, cleaning up the LQP node and making the weak_ptr expired. Maybe easier to discuss this in person tomorrow?

Reply:
Yepp, that's the idea. I'd just run the entire SQLPipeline so that more potential leaks are covered.

Reply:
Adapted the test in 21caa7e.